src/main/scala/nodes/learning/KernelGenerator.scala

```diff
@@ -0,0 +1,273 @@
+        sparseDotProduct(lhsInds, lhsValues, rhsInds, rhsValues)
+      }.toArray)
+    }.toArray
+
+    // val Kblockblock = blockDataArray.map { case (lhsInds, lhsValues) =>
+    //   new DenseVector(blockDataArray.map { case (rhsInds, rhsValues) =>
+    //     sparseDotProduct(lhsInds, lhsValues, rhsInds, rhsValues)
+    //   }.toArray)
+    // }
+
+    Kblock.cache()
+    Kblock.count
+    blockDataBC.unpersist(true)
+
+    (Kblock, KernelUtils.rowsToMatrix(Kblockblock))
+  }
+
+}
+
+class GaussianKernelGenerator(gamma: Double)
+  extends KernelGenerator[DenseVector[Double]]
+  with Serializable {
+
+  def generateKernelTestBlock(
+    testMat: RDD[DenseVector[Double]],
+    trainMat: RDD[DenseVector[Double]],
+    blockIdxs: Seq[Int]): RDD[DenseVector[Double]]= {
+
+    println("testMat.count=" + testMat.count +
+      ",trainMat.count=" + trainMat.count +
+      ", blockSize=" + blockIdxs.length)
+
+    // Dot product of rows of X with each other
+    val trainDotProd = trainMat.map {x =>
+      (x.t*x).toDouble
+    }
+    val testDotProd = testMat.map {x =>
+      (x.t*x).toDouble
+    }
+
+    // b x d block of training data
+    val blockDataArray = trainMat.zipWithIndex.filter{ case (vec, idx) =>
+      blockIdxs.contains(idx)
+    }.map(x=> x._1).collect()
+    val blockData = KernelUtils.rowsToMatrix(blockDataArray)
+    assert(blockData.rows == blockIdxs.length)
+    val blockDataBC = trainMat.context.broadcast(blockData)
+
+    // <xi,xj> for i in [nTest], j in blockIdxs
+    val blockXXT = testMat.mapPartitions { itr  =>
+      val bd = blockDataBC.value
+      val vecMat = KernelUtils.rowsToMatrix(itr)
+      Iterator.single(vecMat*bd.t)
+    }
+    val trainBlockDotProd = DenseVector(trainDotProd.zipWithIndex.filter{ case (vec, idx) =>
+      blockIdxs.contains(idx)
+    }.map(x => x._1).collect())
+    val trainBlockDotProdBC = trainMat.context.broadcast(trainBlockDotProd)
+
+    //R_block = -2.0 * xxt_term + ...
+    //  mat_dot_prod_test * ones(1, length(trainblock)) + ...
+    //  ones(nTest, 1) * mat_dot_prod_train(trainblock)';
+    val kTestBlock = blockXXT.zipPartitions(testDotProd) { case (iterXXT, iterTestDotProds) =>
+      val xxt = iterXXT.next()
+      assert(iterXXT.isEmpty)
+      iterTestDotProds.zipWithIndex.map { case (testDotProd, idx) =>
+        val term1 = xxt(idx, ::).t * (-2.0)
+        val term2 = DenseVector.fill(xxt.cols){testDotProd}
+        val term3 = trainBlockDotProdBC.value
+        val term4 = (term1 + term2 + term3) * (-gamma)
+        exp(term4)
+      }
+    }
+
+    kTestBlock.cache()
+    kTestBlock.count
+    blockDataBC.unpersist()
+    trainBlockDotProdBC.unpersist()
+    kTestBlock
+  }
+
+  // Generate a block of size n x b of the kernel matrix
+  def generateKernelTrainBlock(
+    mat: RDD[DenseVector[Double]],
+    blockIdxs: Seq[Int]): (RDD[DenseVector[Double]], DenseMatrix[Double]) = {
+
+    assert(KernelUtils.isSorted(blockIdxs)) // limitation for now
```

Comment:
Hmm it seems you can only index RowPartitionedMatrices by Range objects,
which have to be sorted...

On Sat, Mar 12, 2016 at 5:26 PM, Vaishaal Shankar <vaishaal@berkeley.edu>
wrote:

> Oh ok lets just have a set, since I convert it to a set anyway.
>
> On Sat, Mar 12, 2016 at 5:23 PM, Stephen Tu <notifications@github.com>
> wrote:
>
>> In src/main/scala/nodes/learning/KernelGenerator.scala
>> <https://github.com/amplab/keystone/pull/234#discussion_r55930609>:
>>
>> > +      }
>> > +    }
>> > +
>> > +    kTestBlock.cache()
>> > +    kTestBlock.count
>> > +    blockDataBC.unpersist()
>> > +    trainBlockDotProdBC.unpersist()
>> > +    kTestBlock
>> > +  }
>> > +
>> > +  // Generate a block of size n x b of the kernel matrix
>> > +  def generateKernelTrainBlock(
>> > +    mat: RDD[DenseVector[Double]],
>> > +    blockIdxs: Seq[Int]): (RDD[DenseVector[Double]], DenseMatrix[Double]) = {
>> > +
>> > +    assert(KernelUtils.isSorted(blockIdxs)) // limitation for now
>>
>> I think this simplifies having one more lookup table-- there is really no
>> reason to prefer any particular ordering of columns within a block, so you
>> might as well take the sorted order. Another way to say this is that
>> blockIdxs should be Set[Int] instead of Seq[Int].
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/amplab/keystone/pull/234/files#r55930609>.
>>
>
>